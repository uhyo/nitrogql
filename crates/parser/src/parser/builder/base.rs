use super::{Rule, utils::PairExt};
use nitrogql_ast::variable::Variable;
use pest::iterators::Pair;

pub fn build_variable(pair: Pair<Rule>) -> Variable {
    let position = pair.to_pos();
    let name = pair.only_child();
    Variable {
        name: name.as_str(),
        position,
    }
}
