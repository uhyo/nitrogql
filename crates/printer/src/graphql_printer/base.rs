use nitrogql_ast::{
    base::Ident,
    r#type::Type,
    value::{StringValue, Value},
};
use sourcemap_writer::SourceMapWriter;

use super::GraphQLPrinter;
use super::utils::print_string;

impl GraphQLPrinter for Type<'_> {
    fn print_graphql(&self, writer: &mut impl SourceMapWriter) {
        match self {
            Type::Named(name) => {
                writer.write_for(name.name.name, self);
            }
            Type::NonNull(non_null) => {
                non_null.r#type.print_graphql(writer);
                writer.write("!");
            }
            Type::List(list) => {
                writer.write("[");
                list.r#type.print_graphql(writer);
                writer.write("]");
            }
        }
    }
}

impl GraphQLPrinter for Value<'_> {
    fn print_graphql(&self, writer: &mut impl SourceMapWriter) {
        match self {
            Value::BooleanValue(value) => {
                writer.write(value.keyword);
            }
            Value::IntValue(value) => {
                writer.write(value.value);
            }
            Value::FloatValue(value) => {
                writer.write(value.value);
            }
            Value::StringValue(value) => {
                value.print_graphql(writer);
            }
            Value::EnumValue(value) => {
                writer.write_for(value.value, self);
            }
            Value::NullValue(value) => {
                writer.write(value.keyword);
            }
            Value::Variable(value) => {
                value.print_graphql(writer);
            }
            Value::ListValue(value) => {
                writer.write("[");
                for (idx, v) in value.values.iter().enumerate() {
                    if idx > 0 {
                        writer.write(",");
                    }
                    v.print_graphql(writer);
                }
                writer.write("]");
            }
            Value::ObjectValue(value) => {
                writer.write("{");
                if value.fields.len() < 2 {
                    for (name, value) in value.fields.iter() {
                        name.print_graphql(writer);
                        writer.write(": ");
                        value.print_graphql(writer);
                    }
                } else {
                    writer.write("\n");
                    writer.indent();
                    for (name, value) in value.fields.iter() {
                        name.print_graphql(writer);
                        writer.write(": ");
                        value.print_graphql(writer);
                        writer.write("\n");
                    }
                    writer.dedent();
                }
                writer.write("}");
            }
        }
    }
}

impl GraphQLPrinter for StringValue {
    fn print_graphql(&self, writer: &mut impl SourceMapWriter) {
        print_string(&self.value, writer);
    }
}

impl GraphQLPrinter for Ident<'_> {
    fn print_graphql(&self, writer: &mut impl SourceMapWriter) {
        writer.write_for(self.name, self);
    }
}
