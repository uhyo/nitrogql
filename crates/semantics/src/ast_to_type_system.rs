use std::borrow::Cow;

use graphql_type_system::{
    DirectiveDefinition, EnumDefinition, EnumMember, Field, InputObjectDefinition, InputValue,
    InterfaceDefinition, Node, ObjectDefinition, ScalarDefinition, Schema, SchemaBuilder,
    TypeDefinition, UnionDefinition,
};
use nitrogql_ast::{
    base::{HasPos, Pos},
    directive::Directive,
    operation::OperationType,
    type_system::{
        ArgumentsDefinition, DirectiveDefinition as AstDirectiveDefinition, FieldDefinition,
        SchemaDefinition, TypeDefinition as AstTypeDefinition, TypeSystemDefinition,
    },
    value::{StringValue, Value},
    TypeSystemDocument,
};

use crate::type_system_utils::{convert_type, ident_to_node};

/// Convert TypeSystemDocument AST to type system struct.
pub fn ast_to_type_system<'src>(
    document: &TypeSystemDocument<'src>,
) -> Schema<Cow<'src, str>, Pos> {
    let mut builder = SchemaBuilder::<Cow<'src, str>, Pos>::new();
    for def in document.definitions.iter() {
        match def {
            TypeSystemDefinition::SchemaDefinition(def) => {
                convert_schema_definition(def, &mut builder);
            }
            TypeSystemDefinition::TypeDefinition(def) => {
                convert_type_definition(def, &mut builder);
            }
            TypeSystemDefinition::DirectiveDefinition(def) => {
                convert_directive_definition(def, &mut builder);
            }
        }
    }

    builder.into()
}

fn convert_schema_definition<'src>(
    def: &SchemaDefinition<'src>,
    builder: &mut SchemaBuilder<Cow<'src, str>, Pos>,
) {
    if let Some(ref desc) = def.description {
        builder.set_description(Node::from(desc.value.clone(), desc.position));
    }
    let root_types = builder.set_root_types(def.position);
    for (operation, def) in def.definitions.iter() {
        match operation {
            OperationType::Query => root_types.set_query_type(ident_to_node(def)),
            OperationType::Mutation => root_types.set_mutation_type(ident_to_node(def)),
            OperationType::Subscription => root_types.set_subscription_type(ident_to_node(def)),
        }
    }
}

fn convert_type_definition<'src>(
    def: &AstTypeDefinition<'src>,
    builder: &mut SchemaBuilder<Cow<'src, str>, Pos>,
) {
    match def {
        AstTypeDefinition::Scalar(def) => builder
            .extend::<Vec<(_, Node<TypeDefinition<_, _>, _>)>>(vec![(
                def.name.name.into(),
                Node::from(
                    TypeDefinition::Scalar(ScalarDefinition {
                        name: ident_to_node(&def.name),
                        description: convert_description(&def.description),
                    }),
                    def.position,
                ),
            )]),
        AstTypeDefinition::Object(def) => builder
            .extend::<Vec<(_, Node<TypeDefinition<_, _>, _>)>>(vec![(
                def.name.name.into(),
                Node::from(
                    TypeDefinition::Object(ObjectDefinition {
                        name: ident_to_node(&def.name),
                        description: convert_description(&def.description),
                        fields: def.fields.iter().map(convert_field).collect(),
                        interfaces: def.implements.iter().map(ident_to_node).collect(),
                    }),
                    def.position,
                ),
            )]),
        AstTypeDefinition::Interface(def) => builder
            .extend::<Vec<(_, Node<TypeDefinition<_, _>, _>)>>(vec![(
                def.name.name.into(),
                Node::from(
                    TypeDefinition::Interface(InterfaceDefinition {
                        name: ident_to_node(&def.name),
                        description: convert_description(&def.description),
                        fields: def.fields.iter().map(convert_field).collect(),
                        interfaces: def.implements.iter().map(ident_to_node).collect(),
                    }),
                    def.position,
                ),
            )]),
        AstTypeDefinition::Union(def) => builder
            .extend::<Vec<(_, Node<TypeDefinition<_, _>, _>)>>(vec![(
                def.name.name.into(),
                Node::from(
                    TypeDefinition::Union(UnionDefinition {
                        name: ident_to_node(&def.name),
                        description: convert_description(&def.description),
                        possible_types: def.members.iter().map(ident_to_node).collect(),
                    }),
                    def.position,
                ),
            )]),
        AstTypeDefinition::Enum(def) => builder
            .extend::<Vec<(_, Node<TypeDefinition<_, _>, _>)>>(vec![(
                def.name.name.into(),
                Node::from(
                    TypeDefinition::Enum(EnumDefinition {
                        name: ident_to_node(&def.name),
                        description: convert_description(&def.description),
                        members: def
                            .values
                            .iter()
                            .map(|mem| EnumMember {
                                name: ident_to_node(&mem.name),
                                description: convert_description(&mem.description),
                                deprecation: convert_deprecation(&mem.directives),
                            })
                            .collect(),
                    }),
                    def.position,
                ),
            )]),
        AstTypeDefinition::InputObject(def) => {
            builder.extend::<Vec<(_, Node<TypeDefinition<_, _>, _>)>>(vec![(
                def.name.name.into(),
                Node::from(
                    TypeDefinition::InputObject(InputObjectDefinition {
                        name: ident_to_node(&def.name),
                        description: convert_description(&def.description),
                        fields: def
                            .fields
                            .iter()
                            .map(|input| InputValue {
                                name: ident_to_node(&input.name),
                                description: convert_description(&input.description),
                                r#type: convert_type(&input.r#type),
                                default_value: input.default_value.as_ref().map(|value| {
                                    // TODO: do not leak
                                    let value_disp = value.to_string().into_boxed_str();
                                    let value_disp = Box::leak(value_disp);
                                    Node::from(&*value_disp, *value.position())
                                }),
                                deprecation: convert_deprecation(&input.directives),
                            })
                            .collect(),
                    }),
                    def.position,
                ),
            )])
        }
    }
}

fn convert_directive_definition<'str>(
    def: &AstDirectiveDefinition<'str>,
    builder: &mut SchemaBuilder<Cow<'str, str>, Pos>,
) {
    builder.extend::<Vec<(_, Node<DirectiveDefinition<_, _>, _>)>>(vec![(
        def.name.name.into(),
        Node::from(
            DirectiveDefinition {
                name: ident_to_node(&def.name),
                description: convert_description(&def.description),
                locations: def
                    .locations
                    .iter()
                    .map(|loc| Node::from(loc.name, loc.position))
                    .collect(),
                arguments: convert_arguments(&def.arguments),
                repeatable: def.repeatable.map(|ident| Node::from((), ident.position)),
            },
            def.position,
        ),
    )])
}

fn convert_description(description: &Option<StringValue>) -> Option<Node<Cow<'static, str>, Pos>> {
    description
        .as_ref()
        .map(|desc| Node::from(desc.value.clone(), desc.position))
}

fn convert_arguments<'src>(
    arguments: &Option<ArgumentsDefinition<'src>>,
) -> Vec<InputValue<Cow<'src, str>, Pos>> {
    arguments.as_ref().map_or(vec![], |args| {
        args.input_values
            .iter()
            .map(|input| InputValue {
                name: ident_to_node(&input.name),
                description: convert_description(&input.description),
                r#type: convert_type(&input.r#type),
                default_value: input.default_value.as_ref().map(|value| {
                    // TODO: do not leak
                    let value_disp = value.to_string().into_boxed_str();
                    let value_disp = Box::leak(value_disp);
                    Node::from(&*value_disp, *value.position())
                }),
                deprecation: convert_deprecation(&input.directives),
            })
            .collect()
    })
}

fn convert_field<'src>(field: &FieldDefinition<'src>) -> Field<Cow<'src, str>, Pos> {
    Field {
        name: ident_to_node(&field.name),
        description: convert_description(&field.description),
        r#type: convert_type(&field.r#type),
        arguments: convert_arguments(&field.arguments),
        deprecation: convert_deprecation(&field.directives),
    }
}

fn convert_deprecation<'src>(directives: &[Directive<'src>]) -> Option<Cow<'src, str>> {
    directives
        .iter()
        .find(|dir| dir.name.name == "deprecated")
        .map(|dir| {
            dir.arguments
                .iter()
                .flat_map(|args| args.arguments.iter())
                .find(|(name, _)| name.name == "reason")
                .map(|(_, value)| value)
                .and_then(|value| match value {
                    Value::StringValue(string) => Some(Cow::Owned(string.value.clone())),
                    _ => None,
                })
                .unwrap_or(
                    // Default value is from the spec
                    Cow::Borrowed("No longer supported"),
                )
        })
}
