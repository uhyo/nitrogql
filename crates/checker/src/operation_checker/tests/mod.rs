use std::borrow::Cow;

use graphql_builtins::generate_builtins;
use graphql_type_system::Schema;
use nitrogql_ast::{base::Pos, OperationDocumentExt, TypeSystemDocument};
use nitrogql_parser::parse_type_system_document;
use nitrogql_semantics::{resolve_operation_extensions, resolve_schema_extensions};

use crate::{check_operation_document, CheckError, OperationCheckContext};

mod operations {
    use std::borrow::Cow;

    use graphql_type_system::Schema;
    use insta::assert_debug_snapshot;
    use nitrogql_semantics::ast_to_type_system;

    use nitrogql_ast::base::Pos;
    use nitrogql_parser::parse_operation_document;

    use super::{parse_to_type_system_document, test_check};

    fn type_system() -> Schema<Cow<'static, str>, Pos> {
        let doc = parse_to_type_system_document(
            "
            schema {
                subscription: S
            }

            type S {
                foo: [Foo!]!
                bar: [String!]!
            }
            type Foo {
                hoge: Int!
                piyo: Int!
            }
        ",
        );
        ast_to_type_system(&doc)
    }

    #[test]
    fn subscription_root_field() {
        let schema = type_system();
        let doc = parse_operation_document(
            "
            subscription a { foo { hoge piyo } }
            subscription b { foo { hoge } bar }
        ",
        )
        .unwrap();

        assert_debug_snapshot!(test_check(schema, doc));
    }

    #[test]
    fn subscription_root_field_recursing() {
        let schema = type_system();
        let doc = parse_operation_document(
            "
            subscription { ...F }
            fragment F on S {
                ...F
            }
        ",
        )
        .unwrap();

        assert_debug_snapshot!(test_check(schema, doc));
    }
}

mod operation_directives {
    use std::borrow::Cow;

    use graphql_type_system::Schema;
    use insta::assert_debug_snapshot;
    use nitrogql_semantics::ast_to_type_system;

    use nitrogql_ast::base::Pos;
    use nitrogql_parser::parse_operation_document;

    use super::{parse_to_type_system_document, test_check};

    fn type_system() -> Schema<Cow<'static, str>, Pos> {
        let doc = parse_to_type_system_document(
            "
            directive @dir_bool_nonnull(bool: Boolean!) repeatable on QUERY | FIELD
            directive @dir_bool(bool: Boolean) repeatable on MUTATION | FIELD

            type Query {
                foo: Int!
            }
            type Mutation {
                bar: Int!
            }

            directive @dir_input(input: MyInput) on QUERY | FIELD
            input MyInput {
                bool: Boolean
                int: Int!
            }
        ",
        );
        ast_to_type_system(&doc)
    }

    #[test]
    fn unknown_directive() {
        let schema = type_system();
        let doc = parse_operation_document(
            "
            query @unknown_dir {
                foo
            }
        ",
        )
        .unwrap();

        assert_debug_snapshot!(test_check(schema, doc));
    }
    #[test]
    fn wrong_location() {
        let schema = type_system();
        let doc = parse_operation_document(
            "
            query @dir_bool {
                foo
            }
        ",
        )
        .unwrap();

        assert_debug_snapshot!(test_check(schema, doc));
    }

    #[test]
    fn wrong_argument_type() {
        let schema = type_system();
        let doc = parse_operation_document(
            "
            query @dir_bool_nonnull(bool: 3) {
                foo
            }
        ",
        )
        .unwrap();

        assert_debug_snapshot!(test_check(schema, doc));
    }

    #[test]
    fn unknown_argument() {
        let schema = type_system();
        let doc = parse_operation_document(
            "
            query @dir_bool_nonnull(bool: false, another: true) {
                foo
            }
        ",
        )
        .unwrap();

        assert_debug_snapshot!(test_check(schema, doc));
    }

    #[test]
    fn no_argument() {
        let schema = type_system();
        let doc = parse_operation_document(
            "
            query q @dir_bool_nonnull {
                foo
            }
            mutation m
              @dir_bool
              @dir_bool(bool: null)
              @dir_bool(bool: true) {
                bar
            }
        ",
        )
        .unwrap();

        assert_debug_snapshot!(test_check(schema, doc));
    }

    #[test]
    fn variable_check() {
        let schema = type_system();
        let doc = parse_operation_document(
            "
            query q($b: Boolean, $b2: Boolean!, $i: Int!)
                @dir_bool_nonnull(bool: $b)
                @dir_bool_nonnull(bool: $b2)
                @dir_bool_nonnull(bool: $i)
                 {
                foo
                @dir_bool(bool: $b)
                @dir_bool(bool: $b2)
            }
        ",
        )
        .unwrap();

        assert_debug_snapshot!(test_check(schema, doc));
    }

    #[test]
    fn unknown_variable() {
        let schema = type_system();
        let doc = parse_operation_document(
            "
            query q1 @dir_bool_nonnull(bool: $b) {
                foo
                @dir_bool(bool: $b)
            }
            query q2($b: Boolean!) @dir_bool_nonnull(bool: $b2) {
                foo @dir_bool(bool: $b2)
            }
        ",
        )
        .unwrap();

        assert_debug_snapshot!(test_check(schema, doc));
    }

    #[test]
    fn input_missing_field() {
        let schema = type_system();
        let doc = parse_operation_document(
            "
            query
            @dir_input(input: { bool: true })
            {
                foo
            }
        ",
        )
        .unwrap();

        assert_debug_snapshot!(test_check(schema, doc));
    }

    #[test]
    fn input_extra_field() {
        let schema = type_system();
        let doc = parse_operation_document(
            "
            query
            @dir_input(input: { bool: true, int: 3 str: \"foobar\" })
            {
                foo
            }
        ",
        )
        .unwrap();

        assert_debug_snapshot!(test_check(schema, doc));
    }

    #[test]
    fn input_field_type_mismatch() {
        let schema = type_system();
        let doc = parse_operation_document(
            "
            query
            @dir_input(input: { bool: true, int: \"foobar\" })
            {
                foo
            }
        ",
        )
        .unwrap();

        assert_debug_snapshot!(test_check(schema, doc));
    }

    #[test]
    fn input_field_null_literal() {
        let schema = type_system();
        let doc = parse_operation_document(
            "
            query
            @dir_input(input: { bool: null, int: 3 })
            {
                foo
            }
        ",
        )
        .unwrap();

        assert_debug_snapshot!(test_check(schema, doc));
    }
}

mod selection_set {
    use std::borrow::Cow;

    use graphql_type_system::Schema;
    use insta::assert_debug_snapshot;
    use nitrogql_semantics::ast_to_type_system;

    use nitrogql_ast::base::Pos;
    use nitrogql_parser::parse_operation_document;

    use super::{parse_to_type_system_document, test_check};

    fn type_system() -> Schema<Cow<'static, str>, Pos> {
        let doc = parse_to_type_system_document(
            "
            type Query {
                foo: Int!
                user: User
                users(name: String): [User!]!
            }
            type User {
                id: ID!
                name: String!
                age: Int
                pet: Animal
                relations: [UserOrPost!]!
            }
            interface Animal {
                id: ID!
                name: String!
            }
            type Post {
                id: ID!
                title: String!
            }
            union UserOrPost = User | Post
        ",
        );
        ast_to_type_system(&doc)
    }

    #[test]
    fn unknown_selected_field() {
        let schema = type_system();
        let doc = parse_operation_document(
            "
            query
            {
                user { id }
                user2
            }
        ",
        )
        .unwrap();

        assert_debug_snapshot!(test_check(schema, doc));
    }

    #[test]
    fn duplicated_selected_field() {
        let schema = type_system();
        let doc = parse_operation_document(
            "
            query q1 {
                user { id }
                user { id name }
            }
            query q2 {
                user { id }
                alias: user { id name } # ‚Üê this is ok
            }
        ",
        )
        .unwrap();

        assert_debug_snapshot!(test_check(schema, doc));
    }

    #[test]
    fn duplicated_alias() {
        let schema = type_system();
        let doc = parse_operation_document(
            "
            query
            {
                user { id }
                user: foo 
            }
        ",
        )
        .unwrap();

        assert_debug_snapshot!(test_check(schema, doc));
    }

    #[test]
    fn invalid_nested_selection() {
        let schema = type_system();
        let doc = parse_operation_document(
            "
            query {
                foo { value }
            }
        ",
        )
        .unwrap();

        assert_debug_snapshot!(test_check(schema, doc));
    }

    #[test]
    fn check_field_arguments() {
        let schema = type_system();
        let doc = parse_operation_document(
            "
            query {
                users(name: \"uhyo\") { id name age }
                user(arg: 123) { id }
            }
        ",
        )
        .unwrap();

        assert_debug_snapshot!(test_check(schema, doc));
    }

    #[test]
    fn check_field_arguments_variable() {
        let schema = type_system();
        let doc = parse_operation_document(
            "
            query($str: String!, $num: Int!, $maybeNum: Int) {
                users1: users(name: $str) { id }
                users2: users(name: $num) { id }
                users2: users(name: $maybeNum) { id }
            }
        ",
        )
        .unwrap();

        assert_debug_snapshot!(test_check(schema, doc));
    }

    #[test]
    fn must_specify_selection_set_for_non_leaf_field() {
        let schema = type_system();
        let doc = parse_operation_document(
            "
            query {
                foo
                user
                users(name: \"uhyo\")
                user2: user {
                    id
                    name
                    age
                }
                users2: users {
                    id
                    name
                }
            }
        ",
        )
        .unwrap();

        assert_debug_snapshot!(test_check(schema, doc));
    }

    #[test]
    fn supports_typename_metafield() {
        let schema = type_system();
        let doc = parse_operation_document(
            "
            query {
                user {
                    __typename
                    pet {
                        __typename
                    }
                    relations {
                        __typename
                    }
                }
            }
        ",
        )
        .unwrap();

        assert_debug_snapshot!(test_check(schema, doc));
    }
}

mod fragments {
    use std::borrow::Cow;

    use graphql_type_system::Schema;
    use insta::assert_debug_snapshot;
    use nitrogql_semantics::{ast_to_type_system, resolve_operation_extensions};

    use crate::{operation_checker::check_operation_document, OperationCheckContext};
    use nitrogql_ast::base::Pos;
    use nitrogql_parser::parse_operation_document;

    use super::{parse_to_type_system_document, test_check};

    fn type_system() -> Schema<Cow<'static, str>, Pos> {
        let doc = parse_to_type_system_document(
            "
            directive @dir_bool(bool: Boolean!) on FIELD
            scalar CustomScalar
            
            type Query {
                foo: Int!
                user: User
                users(name: String): [User!]!
                hasTitle: HasTitle!
                postOrTag: PostOrTag
                userOrSchedule: UserOrSchedule
            }
            type User {
                id: ID!
                name: String!
                age: Int
                userKind: UserKind
                followers: [User!]!
            }
            enum UserKind { NormalUser PremiumUser }
            input MyInput {
                arg: String! = \"\"
            }

            type Post implements HasTitle {
                id: ID!
                title: String!
                body: String!
            }
            interface HasTitle {
                title: String!
            }
            interface HasLabel {
                label: String!
            }

            type Tag implements HasLabel {
                id: ID!
                label: String!
            }
            type Schedule {
                id: ID!
            }
            union UserOrTag = User | Tag
            union PostOrTag = Post | Tag
            union UserOrSchedule = User | Schedule
        ",
        );
        ast_to_type_system(&doc)
    }

    #[test]
    fn unknown_fragment_target() {
        let schema = type_system();
        let doc = parse_operation_document(
            "
            query { user { id } }
            fragment A on Nothing {
                id
            }
        ",
        )
        .unwrap();

        assert_debug_snapshot!(test_check(schema, doc));
    }

    #[test]
    fn invalid_fragment_target() {
        let schema = type_system();
        let doc = parse_operation_document(
            "
            query { user { id }}
            fragment OnScalar on CustomScalar {
                id
            }
            fragment OnEnum on UserKind {
                id
            }
            fragment OnInput on MyInput {
                arg
            }
        ",
        )
        .unwrap();

        assert_debug_snapshot!(test_check(schema, doc));
    }

    #[test]
    fn wrong_fragment_target_obj() {
        let schema = type_system();
        let doc = parse_operation_document(
            "
            query { user {
                ...F
            }}
            fragment F on Post {
                id
                title
            }
        ",
        )
        .unwrap();

        assert_debug_snapshot!(test_check(schema, doc));
    }

    #[test]
    fn wrong_fragment_target_intf() {
        let schema = type_system();
        let doc = parse_operation_document(
            "
            query { user {
                ...F
            }}
            fragment F on HasTitle {
                title
            }
        ",
        )
        .unwrap();

        assert_debug_snapshot!(test_check(schema, doc));
    }

    #[test]
    fn wrong_fragment_target_union() {
        let schema = type_system();
        let doc = parse_operation_document(
            "
            query { user {
                ...F
            }}
            fragment F on PostOrTag {
                ... on Post {
                    id
                }
                ... on Tag {
                    id
                }
            }
        ",
        )
        .unwrap();

        assert_debug_snapshot!(test_check(schema, doc));
    }

    #[test]
    fn wrong_fragment_target_intf_intf() {
        let schema = type_system();
        let doc = parse_operation_document(
            "
            query { hasTitle {
                ...F
                ...G
                ...H
            }}
            fragment F on HasLabel {
                label
            }
            fragment G on HasTitle {
                title
            }
            fragment H on Post {
                title
            }
        ",
        )
        .unwrap();

        assert_debug_snapshot!(test_check(schema, doc));
    }

    #[test]
    fn wrong_fragment_target_intf_union() {
        let schema = type_system();
        let doc = parse_operation_document(
            "
            query { hasTitle {
                ...F
                ...G
            }}
            fragment F on PostOrTag {
                ... on Post {
                    title
                }
                ... on Tag {
                    label
                }
            }
            fragment G on UserOrPost {
                ... on Post {
                    title
                }
                ... on User {
                    name
                }
            }
        ",
        )
        .unwrap();

        assert_debug_snapshot!(test_check(schema, doc));
    }

    #[test]
    fn wrong_fragment_target_union_union() {
        let schema = type_system();
        let doc = parse_operation_document(
            "
            query { userOrSchedule {
                ...F
                ...G
            }}
            fragment F on PostOrTag {
                ... on Post {
                    title
                }
                ... on Tag {
                    label
                }
            }
            fragment G on UserOrTag {
                ... on Tag {
                    label
                }
                ... on User {
                    name
                }
            }
        ",
        )
        .unwrap();

        assert_debug_snapshot!(test_check(schema, doc));
    }

    #[test]
    fn fragment_variables() {
        let schema = type_system();
        let doc = parse_operation_document(
            "
            query($b1: Boolean!) { user {
                ...F
            }}
            fragment F on User {
                id @dir_bool(bool: $b1)
                name @dir_bool(bool: $b2)
            }
        ",
        )
        .unwrap();

        assert_debug_snapshot!(test_check(schema, doc));
    }

    #[test]
    fn inline_fragment_without_condition() {
        let schema = type_system();
        let doc = parse_operation_document(
            "
            query($b1: Boolean!) { user {
                id
                ... {
                    name
                    age
                }
                ... {
                    something
                }
            }}
        ",
        )
        .unwrap();

        assert_debug_snapshot!(test_check(schema, doc));
    }

    #[test]
    fn wrong_inline_fragment_target_obj() {
        let schema = type_system();
        let doc = parse_operation_document(
            "
            query { user {
                ... on Post {
                    id
                    title
                }
            }}
        ",
        )
        .unwrap();

        assert_debug_snapshot!(test_check(schema, doc));
    }

    #[test]
    fn wrong_inline_fragment_target_intf() {
        let schema = type_system();
        let doc = parse_operation_document(
            "
            query {
                user {
                    ... on HasTitle {
                        title
                    }
                }
                hasTitle {
                    ... on HasTitle {
                        title
                    }
                }
            }
        ",
        )
        .unwrap();

        assert_debug_snapshot!(test_check(schema, doc));
    }

    #[test]
    fn wrong_inline_fragment_target_union() {
        let schema = type_system();
        let doc = parse_operation_document(
            "
            query {
                user {
                    ... on PostOrTag {
                        ... on Post {
                            id
                        }
                        ... on Tag {
                            id
                        }
                    }
                }
            }
        ",
        )
        .unwrap();

        assert_debug_snapshot!(test_check(schema, doc));
    }

    #[test]
    fn fragment_spread_recursion() {
        let schema = type_system();
        let doc = parse_operation_document(
            "
            query {
                user {
                    ...F
                }
            }
            fragment F on User {
                id
                ...F
            }
        ",
        )
        .unwrap();

        assert_debug_snapshot!(test_check(schema, doc));
    }

    #[test]
    fn fragment_spread_indirect_recursion() {
        let schema = type_system();
        let doc = parse_operation_document(
            "
            query {
                user {
                    ...F
                }
            }
            fragment F on User {
                id
                ...G
            }
            fragment G on User {
                age
                ...F
            }
        ",
        )
        .unwrap();

        assert_debug_snapshot!(test_check(schema, doc));
    }

    #[test]
    fn fragment_spread_nested_recursion() {
        let schema = type_system();
        let doc = parse_operation_document(
            "
            query {
                user {
                    ...F
                }
            }
            fragment F on User {
                followers {
                    id
                    ...F
                }
            }
        ",
        )
        .unwrap();

        let (doc, _) = resolve_operation_extensions(doc).unwrap();
        let context = OperationCheckContext::new(&schema);
        assert_debug_snapshot!(check_operation_document(&doc, &context));
    }
}

mod imports {
    use std::borrow::Cow;

    use graphql_type_system::Schema;
    use insta::assert_debug_snapshot;
    use nitrogql_semantics::ast_to_type_system;

    use crate::operation_checker::tests::test_check;
    use nitrogql_ast::base::Pos;
    use nitrogql_parser::parse_operation_document;

    use super::parse_to_type_system_document;

    fn type_system() -> Schema<Cow<'static, str>, Pos> {
        let doc = parse_to_type_system_document(
            "
            directive @dir_bool(bool: Boolean!) on FIELD
            scalar CustomScalar
            
            type Query {
                foo: Int!
                user: User
                users(name: String): [User!]!
                hasTitle: HasTitle!
                postOrTag: PostOrTag
                userOrSchedule: UserOrSchedule
            }
            type User {
                id: ID!
                name: String!
                age: Int
                userKind: UserKind
                followers: [User!]!
            }
            enum UserKind { NormalUser PremiumUser }
            input MyInput {
                arg: String! = \"\"
            }

            type Post implements HasTitle {
                id: ID!
                title: String!
                body: String!
            }
            interface HasTitle {
                title: String!
            }
            interface HasLabel {
                label: String!
            }

            type Tag implements HasLabel {
                id: ID!
                label: String!
            }
            type Schedule {
                id: ID!
            }
            union UserOrTag = User | Tag
            union PostOrTag = Post | Tag
            union UserOrSchedule = User | Schedule
        ",
        );
        ast_to_type_system(&doc)
    }

    #[test]
    fn import() {
        let schema = type_system();
        let doc = parse_operation_document(
            "#import F, \"./other.graphql\"
            query {
                user {
                    ...F
                }
            }
        ",
        )
        .unwrap();

        assert_debug_snapshot!(test_check(schema, doc));
    }
}

fn parse_to_type_system_document(source: &str) -> TypeSystemDocument {
    let mut doc = parse_type_system_document(source).unwrap();
    doc.extend(generate_builtins());
    let doc = resolve_schema_extensions(doc).unwrap();
    doc
}

fn test_check(
    schema: Schema<Cow<'static, str>, Pos>,
    doc: OperationDocumentExt,
) -> Vec<CheckError> {
    let (doc, _) = resolve_operation_extensions(doc).unwrap();
    let context = OperationCheckContext::new(&schema);
    check_operation_document(&doc, &context)
}
