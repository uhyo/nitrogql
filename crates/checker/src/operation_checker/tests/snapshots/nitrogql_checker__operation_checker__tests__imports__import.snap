---
source: crates/checker/src/operation_checker/tests/mod.rs
expression: "test_check(schema, doc)"
---
[
    CheckError {
        position: Pos {
            line: 3,
            column: 23,
            file: 0,
            builtin: false,
        },
        message: UnknownFragment {
            name: "F",
        },
        additional_info: [],
    },
]
